src/framework/audio/internal/audioengine.cpp

```diff
@@ -19,289 +19,147 @@
-}
-
-bool AudioEngine::isInited() const
-{
-    std::shared_ptr<IAudioDriver> drv = driver();
-    if (!drv) {
-        return false;
-    }
-
-    if (!drv->isOpened()) {
-        return false;
-    }
+    m_sequencer = std::make_shared<Sequencer>();
 
-    if (!m_inited) {
-        return false;
-    }
+    m_mixer = std::make_shared<Mixer>();
+    m_mixer->setClock(m_sequencer->clock());
 
-    return true;
-}
+    m_buffer = std::make_shared<AudioBuffer>();
+    m_buffer->setSource(m_mixer);
 
-mu::Ret AudioEngine::init()
-{
-    if (isInited()) {
-        return make_ret(Ret::Code::Ok);
-    }
-
-    int res = m_sl->engine.init(SoLoud::Soloud::CLIP_ROUNDOFF,
-                                SoLoud::Soloud::MUAUDIO,
-                                SoLoud::Soloud::AUTO,
-                                SAMPLE_GRANULARITY, // 1024
-                                2);
-
-    if (res == SoLoud::SO_NO_ERROR) {
-        LOGI() << "success inited audio engine";
-        m_inited = true;
-        m_initChanged.send(m_inited);
-        return make_ret(Ret::Code::Ok);
-    }
-
-    m_inited = false;
-    m_initChanged.send(m_inited);
-
-    Err err = Err::UnknownError;
-    if (SoLoud::INVALID_PARAMETER == res) {
-        err = Err::EngineInvalidParameter;
-    } else if (int(Err::DriverNotFound) == res) {
-        err = Err::DriverNotFound;
-    } else if (int(Err::DriverOpenFailed) == res) {
-        err = Err::DriverOpenFailed;
-    }
-
-    LOGE() << "failed inited audio engine, err: " << int(err);
-    return make_ret(err);
-}
-
-void AudioEngine::deinit()
-{
-    m_sl->engine.deinit();
-    m_inited = false;
-    m_initChanged.send(m_inited);
-}
-
-mu::async::Channel<bool> AudioEngine::initChanged() const
-{
-    return m_initChanged;
-}
+    m_sequencer->audioTrackAdded().onReceive(this, [this](ISequencer::audio_track_t player) {
+        m_mixer->addChannel(player);
+    });
 
-float AudioEngine::sampleRate() const
-{
-    return m_sl->engine.getBackendSamplerate();
+    m_worker = std::make_shared<AudioThread>();
+    m_worker->setAudioBuffer(m_buffer);
 }
 
-IAudioEngine::handle AudioEngine::play(std::shared_ptr<IAudioSource> s, float volume, float pan, bool paused)
+AudioEngine::~AudioEngine()
 {
-    IF_ASSERT_FAILED(s) {
-        return 0;
-    }
-
-    IF_ASSERT_FAILED(isInited()) {
-        return 0;
-    }
-
-    s->setSampleRate(sampleRate());
-
-    SoLoud::AudioSource* sa = s->source();
-    IF_ASSERT_FAILED(sa) {
-        return 0;
-    }
-
-    handle h = m_sl->engine.play(*sa, volume, pan, paused);
-
-    pushMeta(h);
-
-    if (paused) {
-        onPause(h);
-    } else {
-        onPlay(h);
-    }
-
-    return h;
 }
 
-void AudioEngine::seek(handle h, time sec)
+bool AudioEngine::isInited() const
 {
-    m_sl->engine.seek(h, sec);
-    onSeek(h);
+    return m_inited;
 }
 
-void AudioEngine::setPause(handle h, bool paused)
+mu::Ret AudioEngine::init(IAudioDriverPtr driver, uint16_t bufferSize)
```

Comment:
This is a design problem that won't be easy to fix.
The problem is that this solution can only work if threads and shared memory between threads are available.
And it cannot work if threads or shared memory is not available, for example, for cases of working with Web Worker or a separate process.
Because the driver is in one thread, and the audio engine and its parts (sequencer, mixer, etc.) are in another thread.

We can keep it that way only if we accept this limitation (the need for the availability of threads and shared memory between it)